firmware/common2015/modules/CommModule/CommPort.hpp

```diff
@@ -0,0 +1,250 @@
+
+    // Check if an RX/TX callback function has been set for the port.
+    bool hasTXCallback() const { return tx_callback != nullptr; }
+    bool hasRXCallback() const { return rx_callback != nullptr; }
+
+    // Check if the port object is
+    bool Exists() const {
+        if (hasRXCallback() || hasTXCallback()) {
+            return true;
+        } else {
+            return this->is_valid;
+        }
+    }
+
+    // Returns a reference to the TX/RX packet count for modifying
+    unsigned int TXPackets() const { return tx_packets; }
+    unsigned int RXPackets() const { return rx_packets; }
+    unsigned int& TXPackets() { return tx_packets; }
+    unsigned int& RXPackets() { return rx_packets; }
+
+    // Standard display function for a CommPort
+    void PrintPort() const {
+        printf("%2u\t\t%u\t%u\t%s\t\t%s\t\t%s\r\n", Nbr(), RXPackets(),
+               TXPackets(), hasRXCallback() ? "YES" : "NO",
+               hasTXCallback() ? "YES" : "NO", isOpen() ? "OPEN" : "CLOSED");
+
+        Console::Flush();
+    }
+
+protected:
+    // Returns the current packet counts to zero
+    void resetPacketCount() {
+        RXPackets() = 0;
+        TXPackets() = 0;
+    }
+
+    // Returns true if the port can provide an RX callback routine
+    bool isReady() const { return (is_open ? true : hasRXCallback()); }
+
+private:
+    // The number assigned to the port
+    uint8_t nbr;
+
+    // If the port is greater than 0, it is a valid port object
+    bool is_valid;
+
+    // If the port is open, it will also be valid
+    bool is_open;
+
+    // Where each upstream & downstream packet count is stored
+    unsigned int rx_packets;
+    unsigned int tx_packets;
+
+    // the class members that hold the function pointers
+    std::function<T> rx_callback;
+    std::function<T> tx_callback;
+};
+
+// Function for defining how to sort a std::vector of CommPort objects
+template <class T>
+bool PortCompare(const CommPort<T>& a, const CommPort<T>& b) {
+    return a < b;
+}
+
+// Class to manage the available/unavailable ports
+template <class T>
+class CommPorts : CommPort<T> {
+private:
+    typename std::vector<CommPort<T>> ports;
+    typename std::vector<CommPort<T>>::iterator pIt;
+
+    const CommPorts<T>& sort() {
+        std::sort(ports.begin(), ports.end(), PortCompare<T>);
+        return *this;
+    }
+
+    const CommPorts<T>& add(const CommPort<T>& p) {
+        pIt = find(ports.begin(), ports.end(), p);
+
+        if (pIt == ports.end()) {
+            ports.push_back(p);
+        }
+
+        return this->sort();
+    }
+
+public:
+    CommPorts<T> operator+=(const CommPort<T>& p) { return this->add(p); }
+
+    CommPort<T>& operator[](const int portNbr) {
+        pIt = find(ports.begin(), ports.end(), (unsigned int)portNbr);
+
+        if (pIt != ports.end()) {
+            return *&(*pIt);
+        }
+
+        throw std::runtime_error("No port for the given number");
```

Comment:
@jjones646, you may need to change this line.  It used to return a reference to a local variable, so I changed it to throw an exception.  The exception causes the whole program to crash now though.  We either need to find where this method is being called with an invalid parameter and fix it or catch the exception.